community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/local_pubsub.py

```diff
@@ -0,0 +1,195 @@
+NOTE: Implementation assumes that both `<TOPIC>` and `.staging/<TOPIC>` are on the same disk device,
+so it can rely on atomic "move / rename" operation.
+"""
+from typing import Any
+import util
+import json
+from dataclasses import dataclass
+from datetime import datetime
+from pathlib import Path
+import os
+import uuid
+
+import logging
+log = logging.getLogger()
+
+
+@dataclass(frozen=True)
+class Message:
+    id: str
+    created: datetime
+    data: Any
+
+    def to_json(self) -> dict[str, str]:
+        return dict(
+            id=self.id,
+            created=self.created.isoformat(),
+            data=self.data)
+    
+    @classmethod
+    def from_json(cls, data: dict[str, str]) -> 'Message':
+        return cls(
+            id=data['id'],
+            created=datetime.fromisoformat(data['created']),
+            data=data['data'])
+
+class Topic:
+    """
+    Acts as PubSub topic (https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.topics).
+    We can have multiple instances of 
+    """
+    def __init__(self, path: Path, staging: Path) -> None:
+        self._path = path
+        self._staging = staging
+    
+    def _gen_id(self, created: datetime) -> str:
+        ts = created.strftime("%Y_%m_%d-%H_%M_%S")
+        suf = str(uuid.uuid4())[:8]
+        return f"{ts}-{suf}"
+    
+    def publish(self, data: Any) -> None:
+        created = util.now()
+        id = self._gen_id(created)
+        msg = Message(id=id, created=created, data=data)
+        
+        staged = self._staging / msg.id
+        dst = self._path / msg.id
+
+        # Write to stagin area first then perform atomic move
+        # to prevent "reads of partial writes"
+        staged.write_text(json.dumps(msg.to_json()))
+        util.chown_slurm(staged)
+        staged.rename(dst)
+
+
+class Subscription:
+    """
+    Acts as PubSub subscription (https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions)
+    with following settings:
+
+    ```
+    ackDeadlineSeconds = +Inf     # don't resend message that was already being delivered but not acked yet
+    retainAckedMessages = False   # don't persist messages that were already acked 
+    enableMessageOrdering = True  # delivers messages in chronoligical order
+    messageRetentionDuration = +Inf # don't expire messages  
+    deadLetterPolicy = None         # "deadlettering" is disabled, subscriber should take care of any poisonous messages
+    retryPolicy = {                 # NACKed message will be re-delievered after some time
+        minimumBackoff = 30s        # NOTE: Practically there is no timer, but Subscription instance will not try to re-deliver NACKed messages.
+        maximumBackoff = 30s        # Assumes that slurmsync runs every 30+ sec.
+    }
+    ```
+
+    IMPORTANT!: Should only be run as part of slurmsync, 
+    this is our way to ensure that at most one instance exists at a time.
+    There is no concurancy safeguards in place.
```

Comment:
I believe even in happy path suspend could be publishing while sync is running. Can you think through and clarify specifically what functionality poses a thread safety risk. (aka, multiple thread calling pull, or acking)

Reply:
Updated `Subscription` class-doc:

```
    IMPORTANT: Should only be run as part of slurmsync, 
    this is our way to ensure that at most one instance exists at a time.
    There is no concurancy safeguards in place, avoid multithreaded `pull`,
    while multithreaded `ack` & `modify_ack_deadline` are OK.
```